lib/persistent_enum.rb

```diff
@@ -192,6 +175,31 @@ def cache_constants(model, required_members, name_attr: "name", required_attribu
       values
     end
 
+    def upsert_records(model, name_attr, expected_rows)
+      # Not all rows will have the same attributes to upsert: group and upsert by attributes
+      expected_rows.group_by(&:keys).each do |row_attrs, rows|
+        upsert_columns = row_attrs - [name_attr, 'id']
+
+        case model.connection.adapter_name
+        when 'PostgreSQL'
+          model.import(rows, on_duplicate_key_update: { conflict_target: [name_attr], columns: upsert_columns })
+        when 'Mysql2'
+          if upsert_columns.present?
+            model.import(rows, on_duplicate_key_update: upsert_columns)
+          else
+            model.import(rows, on_duplicate_key_ignore: true)
+          end
+        else
+          # No upsert support: use first_or_create optimistically
```

Comment:
This method might be easier to understand if it was obvious that only the upsert case cared about the keys grouping. You could achieve this by making a top level if that pulls out the following into the else branch, but then you're checking your database engines twice.

No conclusion or change necessary here, just took me a moment.